src/Plugin/StrawberryRunnersPostProcessor/OcrPostProcessor.php

```diff
@@ -209,28 +209,91 @@ public function run(\stdClass $io, $context = StrawberryRunnersPostProcessorPlug
       // To be used by miniOCR as id in the form of {nodeuuid}/canvas/{fileuuid}/p{pagenumber}
       $page_number = isset($io->input->{$input_argument}) ? (int) $io->input->{$input_argument} : 1;
 
+      //check if PDF is searchable: has DJVU file the layer TXTz?
+      //pdf2djvu -q --no-metadata -p 2 -j0 -o page2.djv some_file.pdf && djvudump page2.djv |grep TXTz |wc -l
+      //
       setlocale(LC_CTYPE, 'en_US.UTF-8');
-      $execstring = $this->buildExecutableCommand($io);
-      error_log($execstring);
-      if ($execstring) {
+      $execstring_checkSearchable = $this->buildExecutableCommand_checkSearchable($io);
+      error_log($execstring_checkSearchable);
+      if ($execstring_checkSearchable) {
         $backup_locale = setlocale(LC_CTYPE, '0');
         setlocale(LC_CTYPE, $backup_locale);
         // Support UTF-8 commands.
         // @see http://www.php.net/manual/en/function.shell-exec.php#85095
         shell_exec("LANG=en_US.utf-8");
-        $proc_output = $this->proc_execute($execstring, $timeout);
-        if (is_null($proc_output)) {
-          throw new \Exception("Could not execute {$execstring} or timed out");
+        $proc_output_checkS = $this->proc_execute($execstring_checkSearchable, $timeout);
+        if (is_null($proc_output_checkS)) {
+          throw new \Exception("Could not execute {$execstring_checkSearchable} or timed out");
         }
 
-        $miniocr = $this->hOCRtoMiniOCR($proc_output, $page_number);
-        error_log($miniocr);
-        $output = new \stdClass();
-        $output->searchapi = $miniocr;
-        $output->plugin = $miniocr;
-        $io->output = $output;
+        error_log($proc_output_checkS);
 
       }
+
+
+      if ($proc_output_checkS == 1) {
+
+        //if searchable run djvu2hocr
+        //
+        setlocale(LC_CTYPE, 'en_US.UTF-8');
+        $execstring_djvu2hocr = $this->buildExecutableCommand_djvu2hocr($io);
+        error_log($execstring_djvu2hocr);
+        if ($execstring_djvu2hocr) {
+          $backup_locale = setlocale(LC_CTYPE, '0');
+          setlocale(LC_CTYPE, $backup_locale);
+          // Support UTF-8 commands.
+          // @see http://www.php.net/manual/en/function.shell-exec.php#85095
+          shell_exec("LANG=en_US.utf-8");
+          $proc_output = $this->proc_execute($execstring_djvu2hocr, $timeout);
+          if (is_null($proc_output)) {
+            throw new \Exception("Could not execute {$execstring_djvu2hocr} or timed out");
+          }
+
+          //djvu2hocr output uses ocrx_line while tesseract uses ocr_line
+          //
+          $proc_output_mod = str_replace('ocrx_line', 'ocr_line', $proc_output);
+
+          $miniocr = $this->hOCRtoMiniOCR($proc_output_mod, $page_number);
+          error_log($miniocr);
+          $output = new \stdClass();
+          $output->searchapi = $miniocr;
+          $output->plugin = $miniocr;
+          $io->output = $output;
+
+        }
+
+        //Do we have to remove djvu file?
```

Comment:
Our processors should do garbage clean up. Question is, when? e.g if the processor passes the output of a file to another processor, then the next processor should do the cleanup
Other option is to make the file managed but temporary so it will get processes/cleanup eventually by Drupal. Wonder if there is any demo code of other modules we could look to see how they do it?

Reply:
In this case, I think the better way is cleanup temporary file (djvu) just before return from processors, to free disk space as soon as possible.
Probably in other cases we have to adopt a main policy to clean garbage, when output or temporary files can be reused in different places. Anyway, when a processor knows that file has no more uses then processor have to clean it ... or not?

Reply:
yes! The sooner the better in this case